### DISCUSSION TITLE
TestDefaultAccountEv.test_puppet_success

### DISCUSSION THREAD

**Discussion by davewiththenicehat** [evennia/evennia#2322] [Evennia Questions]
In short
Shouldn't evennia.accounts.tests.TestDefaultAccountEv.test_puppet_success need to call puppet_object twice for msg to be called with "You are already puppeting this object."?

Long description:

My server/settings.py has MULTISESSION_MODE = 2

If I were to run 'evennia test --settings settings.py' from my server I receive error

```
Traceback (most recent call last):
  File "/home/dave/muddev/evennia/evennia/accounts/tests.py", line 402, in test_puppet_success
    self.account.msg.assert_called_with("You are already puppeting this object.")
  File "/home/dave/muddev/evenv/lib/python3.8/site-packages/mock/mock.py", line 904, in assert_called_with
    raise AssertionError(error_message)
AssertionError: expected call not found.
Expected: mock('You are already puppeting this object.')
Actual: not called.
```

Shouldn't msg not be called? With multisession mode being 2?
For example if I use command 'evennia test evennia' the error does not occur.

The code that calls the assertion error is:

```python
def test_puppet_success(self):
        self.account.msg = MagicMock()
        with patch("evennia.accounts.accounts._MULTISESSION_MODE", 2):
            self.account.puppet_object(self.session, self.char1)
            self.account.msg.assert_called_with("You are already puppeting this object.")
```
For example I just included this on my imported unit tests in server/test_evennia.py. Where I am working on importing all of evennia unit tests and overloading the very very few methods I have to change to support the changes I have made within my code. Allowing me to get all the unit tests from evennia but still support my games specific changes that are not compatible with built in unit tests.

I have changed test_puppet_success to:

```python
def test_puppet_success(self):
        self.account.msg = MagicMock()
        with patch("evennia.accounts.accounts._MULTISESSION_MODE", 2):
            self.account.puppet_object(self.session, self.char1)
            self.account.puppet_object(self.session, self.char1)  # this is the only change
            self.account.msg.assert_called_with("You are already puppeting this object.")
```
There is no assertion error because, I puppet char1 with the first call and msg was called with the expected text on the second.

Before I attempting updating this an PRing it, I wanted to make certain I am not misunderstanding.
If I am in multisession mode 2, I should have to puppet this character, than attempt to puppet again for msg to be called?
The only reason I ask is patch() should be changing multisession mode to 2. If it is in your settings.py file or not. I however do not have a great understanding the patch function. I also lack the understanding of who that attribute is used for specifically. It is obvious what it does over all. I don't know how it specifically interacts with puppet methods.

Thanks

**Comment by Griatch** [comment]
@davewiththenicehat 

The test is based off the `EvenniaTest` parent, which initiates `self.char1` to already be puppeted by the account. So I think the original is correct. Not sure why running the tests in your gamedir would make a difference.
 `evennia test --settings settings.py` will by default run only tests in your mygame folder. How are you importing the default tests to run with this command? 

PS: I fixed the  formatting for you. Please look at [the intro post](https://github.com/evennia/evennia/discussions/2267) for help with how to properly format Markdown to be readable.

**Reply by davewiththenicehat** [reply]
Thank you, that text was not readable previous.
I did not to learn about marking down.

I import tests by importing the module: `from evennia.objects.tests import *`.
In file `server/test_evennia.py`
From there I can overload the classes within the module, as needed. Mostly just replacing default typeclasses. I have been having the overridden module import the module it is overriding. With the same name even. I did not need to `import as`.
When I run `evennia test --settings settings.py .` the imported and overridden modules both run.
As I make changes I have separate unit tests in each of my main folders for my objects. All I want to do here is make certain my changes do not cause issues that make base functionality non functional.

I followed some of the session and puppet logic. It goes fairly deep.
When I run unit tests from any of my game folders I receive errors. Not just this one error, there are 8 total.
I overrode the erroring tests in my `test_evennia` module and mostly manually puppet Char1 to Self.Account resolved the issue. There were one or two other issues. There are two or three unit tests I did not inherit and will have to later. The errors were something I did not want to look into.
With certainty when those tests were running self.account did not have self.char1 as its puppet. I am guessing this has to do with MULTI_SESSION_MODE being 2.

In case it matters, here is my settings file for my primary test server.
```
# Use the defaults from Evennia unless explicitly overridden
from evennia.settings_default import *

######################################################################
# Evennia base server config
######################################################################

# This is the name of your game. Make it catchy!
SERVERNAME = "daveworld"

# open to the internet: 4000, 4001, 4002
# closed to the internet (internal use): 4005, 4006
TELNET_PORTS = [4090]
WEBSOCKET_CLIENT_PORT = 4092
# Have clients communicate via wss after connecting with https to port 4001.
# Without this, you may get DOMException errors when the browser tries
# to create an insecure websocket from a secure webpage.
WEBSOCKET_CLIENT_URL = "wss://daveworld.MYDOMAIN/ws"
WEBSERVER_PORTS = [(4091, 4095)]
AMP_PORT = 4096

# Optional - security measures limiting interface access
ALLOWED_HOSTS = ["daveworld.MYDOMAIN"]

# uncomment if you want to lock the server down for maintenance.
# LOCKDOWN_MODE = True

# Different Multisession modes allow a player (=account) to connect to the
# game simultaneously with multiple clients (=sessions). In modes 0,1 there is
# only one character created to the same name as the account at first login.
# In modes 2,3 no default character will be created and the MAX_NR_CHARACTERS
# value (below) defines how many characters the default char_create command
# allow per account.
#  0 - single session, one account, one character, when a new session is
#      connected, the old one is disconnected
#  1 - multiple sessions, one account, one character, each session getting
#      the same data
#  2 - multiple sessions, one account, many characters, one session per
#      character (disconnects multiplets)
#  3 - like mode 2, except multiple sessions can puppet one character, each
#      session getting the same data.
MULTISESSION_MODE = 2
# The maximum number of characters allowed by the default ooc char-creation command
MAX_NR_CHARACTERS = 2

######################################################################
# Settings given in secret_settings.py override those in this file.
######################################################################
try:
    from server.conf.secret_settings import *
except ImportError:
    print("secret_settings.py file not found or failed to import.")
```

**Reply by Griatch** [reply]
Ah, yes, the MULTISESSION_MODE=2 is likely the culprit; for auto-puppeting, these tests assume mode 0. They should probably be fixed with the `override_settings` decorator: 

```python
...
from django.tests import override_settings

class TestFoo(EvenniaTest):

     @override_settings(MULTISESSION_MODE=0)
     def test_foo(self):
         ...

```

This will run the test in question as if MULTISESSION_MODE was 0. If that works, it would probably be best to update the core tests with this to make it easier to do things like you are doing (PR is welcome!).

**Reply by Griatch** [reply]
@davewiththenicehat Did this solve the issue?
